README.md

```diff
@@ -1,4 +1,107 @@
+
+Evidence-based Higher Kinded Types (HKTs) is a way of emulating HKTs even without explicit language support. It is described in the paper [Lightweight higher-kinded Polymorphism](https://www.cl.cam.ac.uk/~jdy22/papers/lightweight-higher-kinded-polymorphism.pdf) by Jeremy Yallop and Leo White.
+
+A *datatype* is any class who we want to make into an HKT. Examples include List, Future, Optional, Stream, Observable... To consider a class a datatype it has to implement the interface `Hk<F, A>`, or provide a wrapper class that does. In this interface, the generic parameter `F` is considered the *witness* and the generic parameter `A` the *value*.
+
+A *witness* type is an empty type without constructors. This type is unique globally and exclusively used to identify a single datatype when used as a generic parameter. This type has to be different than the datatype, a datatype should not be a witness of itself. Generic parameters requiring a witness are usually represented by the characters F, G, H, or datatype specific characters like L for the left side of an Either.
+
+A *value* is any data type that doesn't affect the operations on the container. For example, you would not expect a `List` of `Strings` to behave differently than a `List` of `UserDto` for operations like `size()` or `map()`. Generic parameters requiring a value are usually represented by the characters A, B, C, D...
+
+Evidence-based HKTs allow transformation from their higher kind form into their datatype form, and viceversa, in what's called an isomorphism. This isomorphism allows transformation in both directions without information loss.
+
+### Spec
+
+An HKT built with KindedJ follows the following guidelines:
+
+* A witness has to be co-located with its datatype and be easily discoverable. Witness types must be available publicly outside the library. In the cases where the witness is dynamically generated, like with annotation processing or compiler plugins, it's expected that the witness belongs at least to the same package or similar physical unit. Witnesses should not be inheritable or used for multiple datatypes.
+
+* By subtyping we obtain a way of transforming datatypes into higher kinds. A datatype must implement the `Hk` interface for its unique witness. That way any operation that'd require a generic HKT is capable of receiving a datatype.
+
+* The reverse operation from HKT to datatype cannot be completed using subtyping. An *evidence function* that allows transforming from HKT to datatype must be provided. Evidence functions have to be available publicly and globally to be accessed by any other KindedJ user.
+
+### Higher Kinded Type Arities
+
+Generic functions and constructs that consume evidence-based HKTs are only capable of understanding kinds of arity 1. To construct HKTs of arity 2+ it is required to nest Hk into its witness type parameter. When trying to represent `F<G, H, A>` it is required to create an instance in the form of `Hk<Hk<Hk<F, G>, H>, A>`.
```

Comment:
Rather than
"Generic functions and constructs that consume evidence-based HKTs are only capable of understanding kinds of arity 1."
it is more:
"The `Hk` interface provided by KindedJ only encoded kinds of arity 1."